### DISCUSSION TITLE
Problem with short-lived file tokens

### DISCUSSION THREAD

**Discussion by benzler** [pocketbase/pocketbase#6137] [Q&A]
Hi there,  

First off, I’d like to thank you for the amazing work on PocketBase!  

We’re encountering occasional issues with protected files in our setup, and I haven’t been able to pinpoint the cause. Here’s an overview of our setup:  

- We have a table field for files with protection enabled.  
- The view rule for files is set to: `@request.auth.id != null`.  
- We’re using a separate S3 bucket for file storage.  

Here’s the issue:  
I am logged in as admin.
When I view entries in the PocketBase frontend and click on a file to open it in a new tab, it usually works fine—loading the file URL with the `?token=...` parameter. However, about 1% of the time, it seems to generate an invalid token. Instead of opening the file, I receive the following error:  

```
{"code":403,"message":"Insufficient permissions to access the file resource.","data":{}}
```

I’ve noticed the following behavior:   
- If an invalid token is generated, it sometimes persists, and even clicking on other files reuses the same invalid token (though this doesn’t always happen).  
- It is not file dependent, if I open the same file entry 15 mins later, it works fine. 

In the PocketBase logs, I see the following entry for these failed requests:  
![Log Screenshot](https://github.com/user-attachments/assets/c7ce24ea-2596-4be4-8924-c706d896d41d)  

I’ve also experienced this behavior when interacting with PocketBase via the API.  

Any help or insights into this issue would be greatly appreciated!

**Comment by ganigeorgiev** [comment]
I'm not sure how to reproduce it, especially _"I’ve also experienced this behavior when interacting with PocketBase via the API."_ (we have automated tests that confirm the correct behavior for the Web API).

Based on what I could understand from the description I guess it could happen if the dashboard is opened long time enough and the cached file token has expired.

We have auto refresh for the cached file token if it is going to expire in the next 10s but the refresh and token replace is not reactive at the moment, aka. it is possible the file token to have expired in the time between the first load and when you click on the preview link.

I'll consider keeping the cached file token in a store value to fix the above issue with the next release but it is also possible that what you are experiencing to be unrelated to that.

**Reply by ganigeorgiev** [reply]
I've submitted a fix in `master` to renew the file token cache in the UI when the user clicks on the file thumb or download link (_I'll push a new release, including backport for PocketBase <= v0.22.x, sometime later this week_).

There are not enough details to confirm that the above will fix your issue, but I suspect that the problem is the same. Or in other words, make sure that the token you are requesting the file with is not expired (you can explore the token payload with [jwt.io](https://jwt.io/) and check the `exp` claim).

**Reply by ganigeorgiev** [reply]
_The above mentioned UI file token cache issue should be fixed in PocketBase [v0.22.28](https://github.com/pocketbase/pocketbase/releases/tag/v0.22.28) and [v0.23.10](https://github.com/pocketbase/pocketbase/releases/tag/v0.23.10) releases._

**Reply by benzler** [reply]
Thanks @ganigeorgiev for the quick response and fix. I'll continue monitoring the behavior, particularly in relation to API interactions, as this was not addressed by the fix. I also understand that, with the current information, it's difficult to fully reconstruct or diagnose the issue. If the problem occurs again, I'll gather details about the token payload. Is there any other information I should collect in case the issue reappears?

**Reply by ganigeorgiev** [reply]
> Is there any other information I should collect in case the issue reappears?

The token `exp` claim compared against the time when it was actually used (aka. when the request was sent) should be enough at least to rule out expired token usage.
